src/authenticators/uds_authenticator/mod.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright 2020 Contributors to the Parsec project.
+// SPDX-License-Identifier: Apache-2.0
+//! Unix domain socket (UDS) peer credentials authenticator
+//!
+//! The `UDSAuthenticator` uses peer credentials from Unix domain sockets to perform
+//! authentication. As such, it uses the UID/GID to grab a username for the owner of the connecting
+//! process. This is used as the application name.
+
+use super::ApplicationName;
+use super::Authenticate;
+use crate::front::listener::ConnectionMetadata;
+use log::error;
+use parsec_interface::requests::request::RequestAuth;
+use parsec_interface::requests::{ResponseStatus, Result};
+
+#[derive(Copy, Clone, Debug)]
+pub struct UDSAuthenticator;
+
+impl Authenticate for UDSAuthenticator {
+    fn authenticate(
+        &self,
+        _auth: &RequestAuth,
+        meta: Option<ConnectionMetadata>,
+    ) -> Result<ApplicationName> {
+        let meta = match meta {
+            Some(meta) => meta,
+            None => {
+                error!("Authenticator did not receive any metadata; cannot authenticate.");
+                return Err(ResponseStatus::AuthenticationError);
+            }
+        };
+
+        let (uid, _gid) = match meta {
+            ConnectionMetadata::PeerCredentials{uid, gid} => (uid, gid),
+            // TODO: add wildcard pattern when `ConnectionMetadata` has more possibilities.
+        };
+
+        Ok(ApplicationName(uid.to_string()))
+    }
```

Comment:
I thought the authentication was meant to verify that the `RequestAuth` contains a string with that UID, just as a "sanity" check

Reply:
Ah, I wasn't aware. Will fix that. 👍

Reply:
Agreed, I think it should. I didn't follow through every line here and I thought there was some Rusty magic going on. My opinion is that this system should be fully explicit with a well-defined authentication type and expected header contents. The calling process should "declare" its UID in the auth header, and the service should use the peer credential to check that it matches.

Reply:
Yeah, it's definitely a fair point. This is now implemented. 🙂